### PR TITLE
Fetch visible root trail record from database

### PR DESCRIPTION
**Pull request by aschempp**
Rendering visible root trails in a tree is a very special case. Because the user actually does not have read permissions on the given record.

We currently use `DataContainer::getCurrentRecord` to fetch visible trail pages. However, if correct permissions are implemented (e.g. on tl_page), the user would not have access to that page. It would therefore not render this node, and therefore not render any nodes of any trees (because the root records are not readable).

I first tried to add a `$checkPermissions` property to `getCurrentRecord`, but that does not really work with our cache of either records or exception.

### CODE REVIEW

**Comment by Toflar** [comment]
Can you please add your explanation as a comment there too because again, this is something that looks irrational without comment if you just look at the code.

**Review by Toflar — DISMISSED** [review]
I'm fine with that. From an architectural standpoint, however, maybe it would be better to always grant READ access to all parents of an allowed tree node? Apparently this is how Contao is designed ;)

**Comment by aschempp** [comment]
> maybe it would be better to always grant READ access to all parents of an allowed tree node? Apparently this is how Contao is designed ;)

I don't think so. You grant access to a specific part of the tree in the user permissions. A user should not have access to nodes not visible in that tree. The trail view is actually a special case that should not exist and could not be implemented in a REST view (but is helpful to users) 🤷

**Comment by Toflar** [comment]
Well in REST it could be implemented by having an extra resource that does a different check so I guess we're good here 👍

File: core-bundle/contao/drivers/DC_Table.php [hunk 1]
```diff
@@ -4326,11 +4326,25 @@ protected function generateTree($table, $id, $arrPrevNext, $blnHasSorting, $intM
...
+			if (!$currentRecord || $checkIdAllowed ? !\in_array($id, $this->visibleRootTrails) : !\in_array($currentRecord['pid'] ?? null, $this->visibleRootTrails))
+			{
+				return '';
+			}
+		}
+
 		// Return if there is no result
 		if (null === $currentRecord)
 		{
-			$objSessionBag->replace($session);
```
Comment (leofeyer) [inline core-bundle/contao/drivers/DC_Table.php:4332]: Is this a related change?

Comment (aschempp) [reply core-bundle/contao/drivers/DC_Table.php:4332]: it is only related in the sense that I noticed it is completely useless (and therefore I could also do an early-return above). 😆

**Comment by aschempp** [comment]
> Well in REST it could be implemented by having an extra resource that does a different check so I guess we're good here 👍

That's true, something like a resource that only returns the title and ID or so.

**Review by Toflar — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @aschempp.
